homeassistant/components/smartthings/smartapp.py

```diff
@@ -0,0 +1,264 @@
+    # occur. Use hass' implementation instead of the built-in one.
+    dispatcher = Dispatcher(
+        signal_prefix=SIGNAL_SMARTAPP_PREFIX,
+        connect=functools.partial(async_dispatcher_connect, hass),
+        send=functools.partial(async_dispatcher_send, hass))
+    manager = SmartAppManager(URL_SMARTAPP, dispatcher=dispatcher)
+    manager.connect_install(functools.partial(smartapp_install, hass))
+    manager.connect_uninstall(functools.partial(smartapp_uninstall, hass))
+
+    hass.http.register_view(APISmartAppView(manager))
+
+    hass.data[DOMAIN] = {
+        DATA_MANAGER: manager,
+        CONF_INSTANCE_ID: config[CONF_INSTANCE_ID],
+        DATA_BROKERS: {}
+    }
+
+
+async def smartapp_install(hass: HomeAssistantType, req, resp, app):
+    """
+    Handle when a SmartApp is installed by the user into a location.
+
+    Setup subscriptions using the access token SmartThings provided in the
+    event. An explicit subscription is required for each 'capability' in order
+    to receive the related attribute updates.  Finally, create a config entry
+    representing the installation.
+    """
+    from pysmartthings import SmartThings, Subscription, SourceType
+
+    api = SmartThings(async_get_clientsession(hass), req.auth_token)
+    coroutines = []
+    for capability in SUPPORTED_CAPABILITIES:
+        sub = Subscription()
+        sub.installed_app_id = req.installed_app_id
+        sub.location_id = req.location_id
+        sub.source_type = SourceType.CAPABILITY
+        sub.capability = capability
+        coroutines.append(api.create_subscription(sub))
+    await asyncio.gather(*coroutines)
+    _LOGGER.debug("SmartApp '%s' under parent app '%s' was installed.",
+                  req.installed_app_id, app.app_id)
+
+    # Add as job not needed because the current coroutine was invoked
+    # from the dispatcher and is not being awaited.
+    await hass.config_entries.flow.async_init(
+        DOMAIN, context={'source': 'install'},
+        data={
+            CONF_APP_ID: app.app_id,
+            CONF_INSTALLED_APP_ID: req.installed_app_id,
+            CONF_LOCATION_ID: req.location_id
+        })
+
+
+async def smartapp_uninstall(hass: HomeAssistantType, req, resp, app):
+    """
+    Handle when a SmartApp is removed from a location by the user.
+
+    Find and delete the config entry representing the integration.
+    """
+    entry = next((entry for entry in hass.config_entries.async_entries(DOMAIN)
+                  if entry.data.get(CONF_INSTALLED_APP_ID) ==
+                  req.installed_app_id),
+                 None)
+    if entry:
+        _LOGGER.debug("SmartApp '%s' under parent app '%s' was removed.",
+                      req.installed_app_id, app.app_id)
+        # Add as job not needed because the current coroutine was invoked
+        # from the dispatcher and is not being awaited.
+        await hass.config_entries.async_remove(entry.entry_id)
+
+
+class APISmartAppView(HomeAssistantView):
```

Comment:
We should not use an HTTP webview but instead use the webhook component.

Reply:
Addressed in latest push!  Thanks!